### ISSUE TITLE
Upscale issues

### ISSUE THREAD

**Issue by cherrynoize** [hollowaykeanho/Upscaler#63]
Really appreciate someone putting some effort into a CLI version for this. I would love to work on some PRs myself but I just don't have the time right now.

So at least I would like to point out that there's some issues with saving the output image, as by default with param `--input image.jpg` I would get an error for it tries to save to `image.jpg/image-upscaled.jpg` (which is clearly not what's expected). Specifying `--output` per se does not solve the issue but I found that `--input ./image.jpg --output ./image-upscaled.jpg` (and probably omitting output too) works around the issue. Though it does seem to ignore the `--output` param completely.

More importantly though:

![111044966_p0_master1200-wm_rm](https://github.com/hollowaykeanho/Upscaler/assets/104397138/28f609a5-47cd-498c-bbd1-0d9ed3b3c9a4)
![111044966_p0_master1200-wm_rm-upscaled](https://github.com/hollowaykeanho/Upscaler/assets/104397138/4a4fac05-647c-4876-a9e9-dd99644f06fe)

I get this.

Input is:

```
upscaler \
                                                  --model upscayl-ultrasharp \
                                                  --scale 2 \
                                                  --input ./111044966_p0_master1200-wm_rm.jpg \
                                          --output ./111044966_p0_master1200-wm_rm-2x.jpg
```

I would love for this to work properly so I could implement it in my dotfiles to auto upscale wallpapers so I could keep them in a more reasonable format until needed. Maybe it's the model? I'll try different ones now.

Also sorry for making this all a single issue. Maybe I should have made two.

**Comment by cherrynoize** [comment]
I get the same result using `upscayl-realesrgan-x4plus-anime` too.

**Comment by hollowaykeanho** [comment]
> So at least I would like to point out that there's some issues with saving the output image, as by default with param --input image.jpg I would get an error for it tries to save to image.jpg/image-upscaled.jpg (which is clearly not what's expected). Specifying --output per se does not solve the issue but I found that --input ./image.jpg --output ./image-upscaled.jpg (and probably omitting output too) works around the issue. Though it does seem to ignore the --output param completely.

Hi @cherrynoize , sorry for the buggy powershell scripting. As a matter of fact, I am working on it but in a separate project repository (since it's rare to find a tested posixshell+powershell standard library). Appreciate your patience.

FYI, next release would be the last one using Shell script wrapping method as I will personally go into the C & C++ codes to make everything scratch.

----



>![111044966_p0_master1200-wm_rm-upscaled](https://github.com/hollowaykeanho/Upscaler/assets/104397138/4a4fac05-647c-4876-a9e9-dd99644f06fe)
> I get this.

This can be caused by 2 issues:
 
1. The input image has an AI poisoning data in it.
2. Model faulty. You can choose other model. I personally use `upscayl-nmkd-4x-superscale-sd-178000-g` (next release or git clone the repo), `upscayl-remacri`, `upscayl-ultramix_balanced`, and lastly `upscayl-ultrasharp-v2`.

**Comment by cherrynoize** [comment]
That would be great. BTW, I thought I had answered to @hollowaykeanho mentioning I wasn't on Windows so I didn't think any `powershell scripting` would be the issue here.

**Comment by Joly0** [comment]
> That would be great. BTW, I thought I had answered to @hollowaykeanho mentioning I wasn't on Windows so I didn't think any `powershell scripting` would be the issue here.

So you arent running this on windows but on linux? If yes, then this is not an issue for me, but for @hollowaykeanho

**Comment by uq1** [comment]
I am having the same issue with the squares being random in the image. It does not happen on every run, but it seems to happen a lot.

I have tried about 6 models and they all seem to do the same thing, so it does not seem to be a model issue.

Image poisoning is somewhat impossible as these are textures from a 20 year old game.

Could this simply be some issue in the image save function? The squares in the image look ok, just not in the right location in the over-all picture.

**Comment by hollowaykeanho** [comment]
> Could this simply be some issue in the image save function? The squares in the image look ok, just not in the right location in the over-all picture.
>
> Here's another example:
> aurora2_upscaled

Wow. This is definitely the first case I've seen. @cherrynoize, and @uq1, can both of you send me a the original image to be included in the test suite please? I will analyze it.

@Joly0, no worries. Please work on the parallel execution task ya. =) I will triage the case.

Apologize for this the late responsiveness. Am currently involved in a client's project. Will patch the software in and out.

**Comment by uq1** [comment]
No problems, here's the original.

![Uploading aurora2.jpg…]()

**Comment by uq1** [comment]
FYI: I just upscaled the same image again and this time it worked fine. The original was being called through a game, while this time I tried it just from command prompt. Could other processes CPU/GPU usage be affecting it?

**Comment by hollowaykeanho** [comment]
@uq1 , not an issue here. It should lower level like drivers or hardware. I'm incline to former.

![sample-03_upscaled](https://github.com/hollowaykeanho/Upscaler/assets/7259572/009bff22-ae0a-4f09-89dd-56ffc00f63f2)

**Comment by hollowaykeanho** [comment]
Yeah... same case for the Ubuntu (I'm on Debian). It should be related to either model or low level driver before we arrive to hardware (@uq1 , your case is very unlikely a hardware problem). Ubuntu is known for heavy modifications for better user experience.

![sample-04_upscaled](https://github.com/hollowaykeanho/Upscaler/assets/7259572/baa8cab9-c8b2-45da-a9ad-bf63f2c0739b)

**Comment by hollowaykeanho** [comment]
Help wanted to further root cause the problem. In the meantime, I do what I can to provide workaround for this weird bug.

**Comment by DrBrule** [comment]
I'm seeing the same issue but only with certain partially AI-generated video.  I've tried many of the models, and for certain input sources it always makes a mosaic.  The generated AI images are fine looking - I had hoped to improve them 2x, but when I try to upscale them they come out mosaic'ed.

**Comment by PedCoelho** [comment]
For me this happens with videos and only on `--scale 2` but not on `-scale 4`. 
Upscaling model choice didn't seem to matter. 
I'm assuming it's a bug related to the scaling factor.

**Comment by codecrafting-io** [comment]
@hollowaykeanho I also had this scaling issue, which could be a model thing, but using the latest [Upscayl NCNN](https://github.com/upscayl/upscayl-ncnn) I was able to use scale 2 (model upscayl-realesrgan-x4plus-v3). Maybe consider use this version which seem better maintained. The exe path must be updated to use Upscayl version

**Comment by ChristianOConnor** [comment]
> [@hollowaykeanho](https://github.com/hollowaykeanho) I also had this scaling issue, which could be a model thing, but using the latest [Upscayl NCNN](https://github.com/upscayl/upscayl-ncnn) I was able to use scale 2 (model upscayl-realesrgan-x4plus-v3). Maybe consider use this version which seem better maintained. The exe path must be updated to use Upscayl version

I am having this mosaic effect as well when I try to scale by 2. How do I execute your workaround? I downloaded the latest [release](https://github.com/upscayl/upscayl-ncnn/releases/tag/20240601-103425) for the Upscayl NCNN repo. I'm on windows and the windows release download zip just contains some dll files, an exe file, and some build instructions if I wanted to compile it myself. How do I use this model with hollowaykeanho/Upscaler? Or what is the easiest way to execute your workaround on Windows?  

Never mind, I figured it out. I downloaded the Windows [release](https://github.com/upscayl/upscayl-ncnn/releases/tag/20240601-103425) for [upscayl-ncnn](https://github.com/upscayl/upscayl-ncnn), unzipped the package, renamed the exe file to `windows-amd64.exe`, then i replaced the existing `windows-amd64.exe` in this app's repository's `bin` folder with the new `windows-amd64.exe` in the upscayl-ncnn release I just downloaded. I also replaced the `vcomp140.dll` and `vcomp140d.dll` in this app's bin folder with the new `vcomp140.dll` and `vcomp140d.dll` from the unzipped upscayl-ncnn build download. After these changes, the command `./start.cmd --video --model upscayl-realesrgan-x4plus-v3 --scale "2" --parallel 2 --input inputfilename.mp4 --output outputfilename.mp4` seemed to work without producing a mosaic artifact-ed output.
